### PR TITLE
feat: update cloudstorageconfiguration to improve copy accross cross-bucket performance

### LINKED ISSUE #60: Using custom options reduces cross-bucket performance
Adding tracking feature request from: https://github.com/googleapis/google-cloud-java/issues/1343

### PR DESCRIPTION
**Pull request by suraj-qlogic**
Fixes #60

### CODE REVIEW

File: google-cloud-nio/src/test/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemTest.java [hunk 1]
```diff
@@ -334,6 +337,43 @@ public void testDeleteRecursive() throws IOException {
...
+            CloudStorageConfiguration.builder().permitEmptyPathComponents(true).build())) {
+      CloudStorageFileSystem destFileSystem =
+          CloudStorageFileSystem.forBucket(
+              "new-bucket",
+              CloudStorageConfiguration.builder().permitEmptyPathComponents(false).build());
+      assertFalse(sourceFileSystem.provider() == destFileSystem.provider());
+      assertNotEquals(sourceFileSystem.config(), destFileSystem.config());
+      assertEquals("bucket", sourceFileSystem.bucket());
+      assertEquals("new-bucket", destFileSystem.bucket());
+    } catch (IOException e) {
```
Comment (elharo) [inline google-cloud-nio/src/test/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemTest.java:373]: IOException should fail test; i.e. don't catch it

Comment (suraj-qlogic) [reply google-cloud-nio/src/test/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemTest.java]: Done

File: google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystem.java [hunk 2]
```diff
@@ -134,6 +134,18 @@ public static CloudStorageFileSystem forBucket(String bucket) {
...
+   *
+   * @param bucketName name of the bucket to initialize {@code CloudStorageFileSystem} object
+   * @return {@code CloudStorageFileSystem} object with existing provider and config
+   * @see #forBucket(String)
+   */
+  private static CloudStorageFileSystem getExistingCloudStorageConfiguration(String bucketName) {
+    return new CloudStorageFileSystem(
+        cloudStorageFileSystem.provider(), bucketName, cloudStorageFileSystem.config());
+  }
+
```
Comment (dmitry-fa) [inline google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystem.java:148]: Method name `getExistingCloudStorageConfiguration` implies returning *existing configuration*, you return FileSystem and create it each time.

File: google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystem.java [hunk 3]
```diff
@@ -195,6 +211,7 @@ public static CloudStorageFileSystem forBucket(
     }
     this.provider = provider;
     this.config = config;
+    this.cloudStorageFileSystem = this;
   }
```
Comment (dmitry-fa) [inline google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystem.java:216]: `this.cloudStorageFileSystem = this` does not work as you probably expect. Each time a new `CloudStorageFileSystem` object is created the static field `cloudStorageFileSystem` will be overridden.

File: google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystem.java (hunk 2 above)
Comment (suraj-qlogic) [reply google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystem.java]: Done

File: google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystem.java (hunk 3 above)
Comment (suraj-qlogic) [reply google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystem.java]: Done

**Comment by suraj-qlogic** [comment]
@dmitry-fa @frankyn ,PTAL

File: google-cloud-nio/src/test/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemTest.java [hunk 4]
```diff
@@ -334,6 +337,39 @@ public void testDeleteRecursive() throws IOException {
...
+  public void testSameProvider() throws IOException {
+    try (CloudStorageFileSystem sourceFileSystem =
+        CloudStorageFileSystem.forBucket(
+            "bucket",
+            CloudStorageConfiguration.builder().permitEmptyPathComponents(true).build())) {
+      CloudStorageFileSystem destFileSystem =
+          CloudStorageFileSystem.forBucket(
+              "new-bucket",
+              CloudStorageConfiguration.builder().permitEmptyPathComponents(true).build());
+      assertEquals(sourceFileSystem.provider(), destFileSystem.provider());
```
Comment (dmitry-fa) [inline google-cloud-nio/src/test/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemTest.java:350]: I think you can replace `assertEquals` with `assertSame` here

File: google-cloud-nio/src/test/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemTest.java [hunk 5]
```diff
@@ -334,6 +337,39 @@ public void testDeleteRecursive() throws IOException {
...
+  public void testDifferentProvider() throws IOException {
+    try (CloudStorageFileSystem sourceFileSystem =
+        CloudStorageFileSystem.forBucket(
+            "bucket",
+            CloudStorageConfiguration.builder().permitEmptyPathComponents(true).build())) {
+      CloudStorageFileSystem destFileSystem =
+          CloudStorageFileSystem.forBucket(
+              "new-bucket",
+              CloudStorageConfiguration.builder().permitEmptyPathComponents(false).build());
+      assertFalse(sourceFileSystem.provider() == destFileSystem.provider());
```
Comment (dmitry-fa) [inline google-cloud-nio/src/test/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemTest.java:367]: assertFalse -> assertNotSame

Comment (suraj-qlogic) [reply google-cloud-nio/src/test/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemTest.java]: Done

File: google-cloud-nio/src/test/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemTest.java (hunk 4 above)
Comment (suraj-qlogic) [reply google-cloud-nio/src/test/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemTest.java]: Done

**Comment by suraj-qlogic** [comment]
@frankyn ,gentle ping

**Pull request merged**